src/ape/api/transactions.py

```diff
@@ -233,14 +236,52 @@ def decode_logs(
         Returns:
             Iterator[:class:`~ape.types.ContractLog`]
         """
-        if not isinstance(abi, (list, tuple)):
-            abi = [abi]
+        if abi:
+            if not isinstance(abi, (list, tuple)):
+                abi = [abi]
 
-        for event_abi in abi:
-            if not isinstance(event_abi, EventABI):
-                event_abi = event_abi.abi
+            for event_abi in abi:
+                if not isinstance(event_abi, EventABI):
+                    event_abi = event_abi.abi
 
-            yield from self.provider.network.ecosystem.decode_logs(event_abi, self.logs)
+                yield from self.provider.network.ecosystem.decode_logs(event_abi, self.logs)
+        else:
+            # If ABI is not provided, decode all events
+            logs_map: Dict[str, Dict[str, Tuple[EventABI, List[Dict]]]] = {}
+            for log in self.logs:
+                address = log["address"]
+                contract_type = self.chain_manager.contracts.get(address)
```

Comment:
i think this is slow, we better do this:
1. extract all addresses from the logs first
2. get all contracts in parallel
3. build a map of `address → selector → abi`
4. decode logs using the above map